cleanlab/experimental/tf_nlp.py

```diff
@@ -0,0 +1,129 @@
+        * ``lnl.fit(training_ids, train_labels, clf_kwargs={'validation_data': val_dataset})``
+        * ``lnl.score(test_ids, test_labels)``
+
+        References:
+        - https://towardsdatascience.com/scikeras-tutorial-a-multi-input-multi-output-wrapper-for-capsnet-hyperparameter-tuning-with-keras-3127690f7f28
+        - https://www.adriangb.com/scikeras/stable/notebooks/DataTransformers.html#4.-Multiple-inputs
+
+        Parameters
+        ----------
+        train_input : dictionary or pandas DataFrame,
+        Tokenized input data. Must contains the following keys/columns (For a more detailed explanation refers to `Huggingface documentation <https://huggingface.co/docs/transformers/main_classes/tokenizer#transformers.PreTrainedTokenizer.__call__>`):
+        - `input_ids`: a list or tensorflow tensor of token ids to be fed to the model. Shape = (num_sample, sequence_len).
+        - `attention_mask` a list or tensorflow tensor of indices specifying which tokens should be attended to by the model. Shape = (num_sample, sequence_len).
+
+        sequence_len : int,
+        Tokenized sentences length.
+
+        kwargs : optional,
+        Optional arguments useful to fit a `Tensorflow model`. Refers to Tensorflow documentation for a detailed list of attributes `<https://www.tensorflow.org/api_docs/python/tf/keras/Model#fit>`
+        """
+        super().__init__(**kwargs)
+        self.train_input = train_input
+        self.seq_len = seq_len
+
+    def split_input(self, X):
+        splitted_X = [
+            X[:, : self.seq_len],  # input_ids
+            X[:, self.seq_len :],  # attention_mask
+        ]
+        return splitted_X
+
+    @property
+    def feature_encoder(self):
+        return FunctionTransformer(
+            func=self.split_input,
+        )
+
+    def _get_tf_input(self, ids, train_input):
+        indexed_input_ids = np.array(tf.gather(train_input["input_ids"], indices=ids))
+
+        indexed_attention_mask = np.array(tf.gather(train_input["attention_mask"], indices=ids))
+
+        return np.hstack([indexed_input_ids, indexed_attention_mask])
+
+    def fit(self, ids, y, sample_weight=None, **kwargs):
+        """
+        Constructs and fit a new model using the given data.
+        Parameters
+        ----------
+        ids : array-like of shape (n_samples,)
+            Ids of training samples to be used to train the model.
+        y : array-like of shape (n_samples,)
+            True labels.
+        sample_weight : array-like of shape (n_samples,), default=None
+            Array of weights that are assigned to individual samples.
+            If not provided, then each sample is given unit weight.
+        **kwargs : Dict[str, Any]
+            Extra arguments to route to Tensorflow `Model.fit`.
+        """
+        X = self._get_tf_input(ids, self.train_input)
+        return super().fit(X, y, sample_weight=sample_weight, **kwargs)
+
+    def predict_proba(self, ids, **kwargs):
```

Comment:
After user has run `Cleanlearning.fit()`, how are they supposed to use this classifier to produce predictions on *new* test data?  Your example did not demonstrate how that would work.

Reply:
I think `predict()` and `predict_proba()` should look more like this (I've only shown `predict` below but `predict_proba` should look similar). What do you think?

```
def predict_proba(self, ids=None, *, test_input=None, **kwargs):
    if ids is not None:
        if test_input is not None:
            raise ValueError("One of ids or test_input must be None")
        X = self._get_tf_input(ids, self.train_input)
    elif test_input is not None:
        X = self._get_tf_input(np.arange(len(test_input), test_input)  # probably need to replace len(...) with proper way to get length of the test_input tokenizer 
    else:
        raise ValueError("Both ids and test_input cannot be None")
    return super().predict_proba(X, **kwargs)
```

Then presumably your example code could look like this to appear more complete:  
```
cl = Cleanlearning.fit(...)`:
test_preds = cl.predict(val_input)
# report test accuracy of test_preds vs test_df['label']
```

Reply:
Yes, we can manage new data in this way, even if the best thing would be to take the original data as input instead of the indexes. In that way, the code would be better and there would be no need to have the `test_input` extra field. For the time being, we can modify the code as you suggested and fix this issue in future.

Reply:
yes sounds good thanks for making the update!